File: data/analysis.py
import pandas as pd
import numpy as np
from bs4 import BeautifulSoup
import re

# import matplotlib.pyplot as plt

# Question cols
# print(df_q.columns.values)
# ['Id' 'PostTypeId' 'AcceptedAnswerId' 'ParentId' 'CreationDate'
#  'DeletionDate' 'Score' 'ViewCount' 'Body' 'OwnerUserId' 'OwnerDisplayName'
#  'LastEditorUserId' 'LastEditorDisplayName' 'LastEditDate'
#  'LastActivityDate' 'Title' 'Tags' 'AnswerCount' 'CommentCount'
#  'FavoriteCount' 'ClosedDate' 'CommunityOwnedDate']

# Answer cols
# print(df_a.columns.values)
# ['Id' 'PostTypeId' 'AcceptedAnswerId' 'ParentId' 'CreationDate'
#  'DeletionDate' 'Score' 'ViewCount' 'Body' 'OwnerUserId' 'OwnerDisplayName'
#  'LastEditorUserId' 'LastEditorDisplayName' 'LastEditDate'
#  'LastActivityDate' 'Title' 'Tags' 'AnswerCount' 'CommentCount'
#  'FavoriteCount' 'ClosedDate' 'CommunityOwnedDate' 'Id.1' 'PostTypeId.1'
#  'AcceptedAnswerId.1' 'ParentId.1' 'CreationDate.1' 'DeletionDate.1'
#  'Score.1' 'ViewCount.1' 'Body.1' 'OwnerUserId.1' 'OwnerDisplayName.1'
#  'LastEditorUserId.1' 'LastEditorDisplayName.1' 'LastEditDate.1'
#  'LastActivityDate.1' 'Title.1' 'Tags.1' 'AnswerCount.1' 'CommentCount.1'
#  'FavoriteCount.1' 'ClosedDate.1' 'CommunityOwnedDate.1']

# Gotcha: parens needed in conditions
df = pd.read_csv('Questions.csv')
df['Accepted'] = np.where(df['AcceptedAnswerId'].notnull(), 1, 0)
df = df[(df['ClosedDate'].isnull()) & (df['Score'] > 0)]
df = df.loc[:, ['Score', 'Accepted', 'AnswerCount', 'CommentCount']]
# df = df.groupby(['Score']).mean()
# if selecting multiple
# df = df.groupby(['Score', 'Accepted']).mean()
# by : mapping, function, str, or iterable (also don't need "by=")
# the below groupby is like doing ". ~ Score" for R's by in aggregate
df = df.groupby('Score', as_index=False).mean()
# by : str or list of str
df = df.sort_values('Score', ascending=False)
# df.corr('pearson')
# print(df.sort_values('Score', ascending=False))
print(df.corr('pearson'))

# Holds all the questions
df = pd.read_csv('Questions.csv')
df_q = df[df['ClosedDate'].isnull()]
df_qt = len(df_q)
# print("Number of questions total: %d" % df_qt)
# print(df_q[['ClosedDate']])
# In Python we use [] to select columns
# ':' is used to slice
# For vanilla slice we have [n:n], [:n], [:n] [n:n:+/-n]
# all rows with all cols (no slice)
# print(df_q[:])
# first row with all cols
# print(df_q[1:2])
# you can use a 3rd number for steps
# e.g. every other
# print(df_q[0:7:2])
# you can also go in reverse
# print(df_q[6::-1])
# print(df_q[:6:-1])

# Gotcha: you can't use +/- steps for anything other than
# vanilla slicing.
# use loc for label-based indexing 
# all rows all cols
# print(df_q.loc[:,])
# all rows 1 or more cols 
# print(df_q.loc[:,'ClosedDate'].head(100))
# Example of transfer issue within Python, the below code doesn't
# work and prints all rows instead of filtering null values
# Another issue is that usually in Python you use null or not null
# to filter for nulls but here isnull() and notnull() are used.
# print(df['ClosedDate'].dtype)

# gotcha: you need to wrap each condition here with parens
# if you don't you get different results

# you can also have condition for the : part of loc
# print(df.query('ClosedDate.isnull() & Score == 0').head(2))
# print(df.loc[(df['ClosedDate'].isnull()) & (df['Score'] == 0), ['ClosedDate', 'Score']].head(2))
# print(df.loc[(df['ClosedDate'].isnull()) & (df['Score'] == 0), ['ClosedDate', 'Score']].head(2))

# print(len(df_q.query('(AnswerCount > 0) & (Score > 0)')))

# print(df.loc[df['ClosedDate'].isnull() & df['Score'] == 0, ['ClosedDate', 'Score']])
# Equivalent:
# print(df[(df['ClosedDate'].isnull()) & (df['Score'] == 0)].loc[:,['ClosedDate', 'Score']])


# first two rows and all cols
# print(df_q[1:3])

# use ix for indexing dfs based on indices
# first and third row and all cols
# print(df_q.ix[[0,2]])
# using iloc:
# print(df_q.iloc[[0,2],:])

# iloc for
# first row and 2nd and third col
# print(df_q.iloc[0:1,1:3])

# first, second row and second and third col
# print(df_q.iloc[0:2,1:3])

# all rows and first column
# print(df_q.iloc[:,0:1])
# or:
# print(df_q.iloc[:,[0]])

# all rows and first and third col
# print(df_q.iloc[:,[0,1]])



# Holds all the answers
df_a = pd.read_csv('QueryAllAnswers.csv')
df_at = len(df_a)
# print("Number of answers total: %d" % df_at)

# Given a filtered dataframe, print total rows and percentage from total
def print_stats(df, s):
	df_t = len(df)
	df_p = int(((df_t / float(df_qt)) * 100))
	print("Number of questions with %s: %d (%d%%)" % (s, df_t, df_p))

# # Descriptive stats for Table 1 answering RQ1 generally

# # Looking at open questions

# questions with at least an answer
# df_answer = df_q[df_q.AnswerCount > 0]
# print_stats(df_answer, 'answers')

# # questions with no answer
# df_no_answer = df_q[df_q.AnswerCount == 0]
# print_stats(df_no_answer, 'no answers')

# # questions with accepted answer
# df_accepted = df_q[df_q['AcceptedAnswerId'].isnull()]
# print_stats(df_accepted, 'an accepted answer')

#  # questions with score >= 3
# df_vo = df_q[df_q.Score >= 3]
# print_stats(df_vo, 'votes >= 3')

# # questions with views >= 300
# df_vu = df_q[df_q.ViewCount >= 300]
# print_stats(df_vu, 'views >= 100')

# # questions with favorites
# df_f = df_q[df_q.FavoriteCount > 0]
# print_stats(df_f, 'favorites')

# # questions with code markup or code sample
# # df_code = df_q[df_q]

# # html_text = """
# # <h2>this is cool #12345678901</h2>
# # <h2>this is nothing</h2>
# # <h2>this is interesting #126666678901</h2>
# # <h2>this is blah #124445678901</h2>
# # """

# # soup = BeautifulSoup(html_text, 'html.parser')

# # pattern = re.compile(r'blah')
# # print(soup.find('h2', text=pattern))

# # Returns a list of ids of df that contains a given html in body
# def tags(tag):
# 	ids = []
# 	for index, row in df_q.iterrows():
# 		soup = BeautifulSoup(row['Body'], 'html.parser')
# 		if len(soup.find_all(tag)) > 0:
# 			ids.append(row['Id'])
# 	return ids

# # HREF LINKS GENERAL

# # questions with links
# links_ids = tags('a')
# df_links = df_q[df_q['Id'].isin(links_ids)]
# df_links_t = df_links.shape[0]
# df_links_p = int(((df_links_t / float(df_qt))*100))
# print("Number of questions with links: %d (%d%%)" % (df_links_t, df_links_p))

# # HREF LINKS SPECIFIC (could be cleaned up but will do for now)
# so_links = []
# droid_links = []
# other_links = []
# for index, row in df_links.iterrows():
# 	soup = BeautifulSoup(row['Body'], 'html.parser')
# 	so_urls = soup.find_all('a', href=re.compile('stackoverflow.com'))
# 	droid_urls = soup.find_all('a', href=re.compile('developer.android.com'))
# 	if len(droid_urls) > 0:
# 		droid_links.append(row['Id'])
# 	elif len(so_urls) > 0:
# 		so_links.append(row['Id'])
# 	else:
# 		other_links.append(row['Id'])

# # questions with links to other SO posts: stackoverflow.com
# df_so = df_links[df_links['Id'].isin(so_links)]
# df_so_t = df_so.shape[0]
# df_so_p = (df_so_t / float(df_links_t))*100
# print("Number of questions with SO links: %d (%.2f%%)" % (df_so_t, df_so_p))

# # questions with links to Android Documentation: developer.android.com
# df_d = df_links[df_links['Id'].isin(droid_links)]
# df_d_t = df_d.shape[0]
# df_d_p = (df_d_t / float(df_links_t))*100
# print("Number of questions with Android links: %d (%.2f%%)" % (df_d_t, df_d_p))

# # questions with links to other sites
# df_o = df_links[df_links['Id'].isin(other_links)]
# df_o_t = df_o.shape[0]
# df_o_p = (df_o_t / float(df_links_t))*100
# print("Number of questions with other links: %d (%.2f%%)" % (df_o_t, df_o_p))

# # CODE LINKS (either markup for single word or code sample)

# # questions with code markup or code sample
# codes_ids = tags('code')
# df_codes = df_q[df_q['Id'].isin(codes_ids)]
# df_codes_t = df_codes.shape[0]
# df_codes_p = int(((df_codes_t / float(df_qt))*100))
# print("Number of questions with codes: %d (%d%%)" % (df_codes_t, df_codes_p))

# # questions with single word code markup

# # questions with code sample

# # randomly sample 300 questions that have answers
# # df_r = df_q.sample(300)
# # with open('random_sample.html', 'w') as f:
# # 	f.write('<!DOCTYPE html>\n')
# # 	f.write('<html lang="en-US">\n')
# # 	f.write('<head><style>body {background-color: #5c8457;}' \
# # 		+ ' h1 {color: white;}p, a {color: #E8AA0C;}</style></head>')
# # 	f.write('<body><p>\n')
# # 	values = list(df_r['Id'].values)
# # 	# print(values)
# # 	for i in range(0, len(values)):
# # 		f.write('<a href=\"https://stackoverflow.com/questions/' + \
# # 			str(values[i]) + '\" target=\"_blank\">' + str(values[i])+'</a>\n')
# # 	f.write('</p><body>\n')
# # 	f.write('</html>')

# # TODO: Descriptive stats for Table 1 answering RQ1 by category on random 
# # subset of questions

File: data/analysis.r
# R version of analysis.py
library(plyr)
library(tibble)
library(dplyr)
library(lubridate)
require(stats)

# sprintf is like print() in Python3 EXCEPT 
# you use comma separator for values in R

# the df[col] syntax does not work in R
# instead use df$col to index col

# a gotcha: when subsetting a df, remember to specify the indices
#   for the col as well df[df$col == '', ]
# this is like Python's slicing syntax using : like so df[]
# If you want to avoid the gotcha, you can also use subset(), which
# will allow you to simply pass in df and condition for 2nd arg

# Flexibility of assignment in R
# Holds all the questions
df <- read.csv('QueryQuestions.csv')
# read.csv('QueryQuestions.csv') -> df
# gotcha: be sure to surround variable name in quotes for assign
# assign("df", read.csv('QueryQuestions.csv'))

# Next confusion is regarding:
#   data type assumption
#   subset by variable name 
#   subset by two conditions (confusing with &)

# Note the flexibility of subsetting in R, you can use 3 versions:
#   df[df[,'col'] operator value, ]
#   df[df$col operator value, ]
#   subset(df, col operator value)

# In Python ClosedDate without data was NA, in R it is '', it's hard knowing
# when empty values will turn into NA or ''
# Use head to peek at rows from top
# head(df[, 'ClosedDate'], 100)
# Mistake: using is.na
# df_q <- subset(df, is.na(ClosedDate))
# df_q <- subset(df, ClosedDate == '')
# Equivalents:
# df_q <- df[df[, 'ClosedDate'] == '', ]
# df_q <- df[df$ClosedDate == '',]

# Logical operator for and is & not &&
# Als note here that = is used for select not <- or ->
# df_q <- subset(df, ClosedDate == '' & Score == 0, 
#     select = c('ClosedDate', 'Score'))
# Equivalents:
# Note here how () is NOT needed for conditions
# df_q <- df[df$ClosedDate == '' & df$Score == 0, c('ClosedDate', 'Score')]
# df_q <- df[which(df$ClosedDate == '' & df$Score > 0),]
# select(df, 1:10)[1:2,]
df <- read.csv('QueryQuestions.csv')
df_q <- subset(df, ClosedDate == '' & AnswerCount > 0 & Score > 0, 
    select = c('ClosedDate', 'Score'))

# Base R for filter
df_q <- df[df$ClosedDate == '' & df$AnswerCount > 0 & df$Score > 0,]
# dplyr filter
# Can use comma for & or just &, | for or
# Note that nothing is modified in-place if not assigned to var
df_q <- filter(df, ClosedDate == '', AnswerCount > 0, Score > 0)
# Another way
# df_q <- filter(df, ClosedDate == '' & AnswerCount > 0 & Score > 0)

# Base R for select (hard to write)
df_q <- df_q[ ,c('AnswerCount', 'Score')]
# dplyr select, easier; it even looks more Pythonic :)
df_q <- select(df_q, AnswerCount, Score, Title)
# fancier one that uses condition of selecting cols
# df_q <- select(df_q, AnswerCount, Score, contains("Date"))
df_q[['ClosedDate']]
# Base R arrange/order
# df_q <- df_q[]

fx <- function(x) {
    return(ifelse(!is.na(x),1,0))
}

# Base R approach
df <- read.csv('QueryQuestions.csv')
# c <- 1:dim(df)[1]
df <- df[df$ClosedDate == '' & df$Score > 0, ]
# most basic way of populating a new column with binary value
df$Accepted <- unlist(ifelse(!is.na(df$AcceptedAnswerId), 1, 0))
# this has the same effect as above
# df$Accepted <- Vectorize(ifelse(!is.na(df$AcceptedAnswerId), 1, 0))
# another way is to use lapply which returns a list, apply unlist to vectorize
# df$Accepted <- unlist(lapply(df$AcceptedAnswerId, function(x) ifelse(!is.na(x), 1, 0)))
# a better way if using an apply() is to use sapply() which vectorizes results
# df$Accepted <- sapply(df$AcceptedAnswerId, function(x) ifelse(!is.na(x), 1, 0))
# cols <- 
df <- df[, c("Score", "Accepted", "ViewCount", "AnswerCount", "CommentCount")]
# trick/gotcha: you don't even need the data=df_q OR FUN=mean
# gotcha: you need to caps the FUN
# gotcha: you need to make data all lower
# trick: add a - operator for desc instead of decreasing=TRUE
df <- aggregate(. ~ Score, data=df, FUN=mean, na.rm=TRUE)
df <- df[order(-df$Score), ]
cor(df, use="complete.obs", method="pearson")
print(head(df, 10))


# Tidy-verse approach
df_t <- read.csv('QueryQuestions.csv') %>% 
    filter(ClosedDate == '', Score > 0) %>%
    mutate(Accepted = ifelse(!is.na(AcceptedAnswerId), 1, 0)) %>%
    select(Score, Accepted, ViewCount, AnswerCount, CommentCount) %>%
    group_by(Score) %>%
    # summarise_at(funs(mean), list("Accepted", "ViewCount", "AnswerCount", "CommentCount"))
    summarise_all(funs(mean)) %>%
    # This is how you'd do it if renaming and picking what funs to apply
    # summarise(Accepted = mean(Accepted), 
    #     ViewCount = mean(ViewCount), 
    #     AnswerCount = mean(AnswerCount), 
    #     CommentCount = mean(CommentCount)) %>%
    arrange(desc(Score))
# Look at correlation btw two things (Pearson default)
# cor(df_q$Score, df_q$Accepted)
cor(df_t, use="complete.obs", method="pearson") 
print(head(df_t, 10))
# Forces tibble into a dataframe when printing
# data.frame(head(df_q,25))

# print(head(df_q, 100))
# R subset df: 
#   format -> 
# df[num|range|list of cols : num|range|list of cols, +/-step]
# All Rows and All Columns
# df[,]
# # First row and all columns
# df[1,]
# # First two rows and all columns
# df[1:2,]
# Add a comman and a -num to go in reverse 
# df[6:1,1:3,-1]
# # First and third row and all columns
# Gotcha: what happens with row at 0?
# There is no error, it is simply ignored
# df_q[c(1,3), ]
# # First Row and 2nd and third column
# df[1, 2:3]
# # First, Second Row and Second and Third COlumn
# df[1:2, 2:3]
# # Just First Column with All rows
# df[, 1]
# # First and Third Column with All rows
# df[,c(1,3)]


# Gotcha: In Python df[1:4] slices and prints result but in R it prints all
# rows with the first 4 columns. 
# Gotcha: Also note how 1:4 is inclusive in R, meaning we count 4 items but 
# in Python 4 is exclusive. This is bc it's 1-based whereas Python is 0-based.
# all rows with 4 cols
# print(df_q[1:4,1:4])
# 4 rows with all cols
# print(df_q[1:4,])

# specifying cols gives us values for the specific column
# Gotcha: typeof
# note how the typeof for the result here is actual typeof of the col, not the
# result of the subsetting! So typeof is more about the element of the result.
# Above, typeof(df_q[1:4,]) would give us list bc each element is a list of 
# values by col
# print(df_q[1:2000,1])


# select(df_q[df_q$Score == 0 && df_q,], CreationDate, Title)
# Subset according to list of columns
# df_q[,c("CreationDate", "Title", "Score", "AnswerCount")][1:5,]

# Holds all the answers
df_a <- read.csv('QueryAllAnswers.csv')
df_at <- nrow(df_a)
# sprintf("Number of answers total: %d", df_at)

# Function defintion differs from Python 
# Notice the functional-style of programming
# function has to be assigned to a variable, a function is
# just like a variable, although the same can be done in
# Python, it's not enforced as functions are not first-class
# by default (programmers are more used to using def by itself)
# Notice how the s is assigned with '=' this is bc in this case
# <- is not allowed and now it makes sense when to use which.
#   You will get an unexpected assignment error if you try <-
ps <- function(df, s="some stat") {
    # Flexibility: notice how the data type for df_qt didn't have to
    # be converted to a float in order to get the decimal points for
    # pct
    sprintf("Number of questions with %s: %d (%.2f%%)", 
        s, 
        nrow(df), 
        (nrow(df) / df_qt) * 100)
}

# questions with at least an answer
df_answer <- df_q[df_q$AnswerCount > 0,]
# ps(df_answer)

# questions with no answer
df_no_answer <- df_q[df_q$AnswerCount == 0,]
# ps(df_no_answer, 'no answers')
